jsettlers.main.swing/src/jsettlers/main/swing/foldertree/FilesystemTreeNode.java

```diff
@@ -0,0 +1,277 @@
+package jsettlers.main.swing.foldertree;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Enumeration;
+import java.util.Iterator;
+import java.util.List;
+
+import javax.swing.tree.TreeNode;
+
+import jsettlers.graphics.swing.resources.SettlerFolderChecker;
+
+/**
+ * Node to display file / folders
+ * 
+ * @author Andreas Butti
+ */
+public class FilesystemTreeNode implements TreeNode {
+
+	/**
+	 * If the children are already loaded
+	 */
+	protected boolean loaded = false;
+
+	/**
+	 * If this is a settler folder
+	 */
+	private boolean settlerFolder = false;
+
+	/**
+	 * To expand the node on the first selection
+	 */
+	private boolean wasExpanded = false;
+
+	/**
+	 * Tmp loaded child list
+	 */
+	private final List<FilesystemTreeNode> list = new ArrayList<>();
+
+	/**
+	 * Current child list
+	 */
+	private List<FilesystemTreeNode> children = new ArrayList<>();
+
+	/**
+	 * Parent node
+	 */
+	private TreeNode parent;
+
+	/**
+	 * The file represented by this node
+	 */
+	private final File file;
+
+	/**
+	 * Constructor
+	 * 
+	 * @param file
+	 *            The file to represent
+	 */
+	public FilesystemTreeNode(File file) {
+		this.file = file;
+	}
+
+	/**
+	 * @return The file
+	 */
+	public File getFile() {
+		return file;
+	}
+
+	/**
+	 * @return Root node
+	 */
+	private RootTreeNode findRoot() {
+		TreeNode node = getParent();
+		while (true) {
+			if (node == null) {
+				return null;
+			}
+			if (node instanceof RootTreeNode) {
+				return (RootTreeNode) node;
+			}
+			node = node.getParent();
+		}
+	}
+
+	/**
+	 * Load the children of this folder, if a folder
+	 */
+	private void loadChildren() {
+		RootTreeNode root = findRoot();
+		if (root == null) {
+			loadChildren1();
+			applyLoadedChildren();
+		} else {
+			root.loadAsynchron(this);
+		}
+	}
+
+	/**
+	 * @return true if this node already was expanded
+	 */
+	public boolean wasExpanded() {
+		return wasExpanded;
+	}
+
+	/**
+	 * @param wasExpanded
+	 *            true if this node already was expanded
+	 */
+	public void setWasExpanded(boolean wasExpanded) {
+		this.wasExpanded = wasExpanded;
+	}
+
+	/**
+	 * @return if this is a settler folder
+	 */
+	public boolean isSettlerFolder() {
+		return settlerFolder;
+	}
+
+	/**
+	 * Load children if not loaded
+	 */
+	private synchronized void loadIfNotLoaded() {
+		if (!loaded) {
+			loaded = true;
+
+			loadChildren();
+		}
+	}
+
+	@Override
+	public boolean isLeaf() {
+		loadIfNotLoaded();
+		if (children != null) {
+			return children.isEmpty();
+		}
+		return true;
+	}
+
+	@Override
+	public int getChildCount() {
+		loadIfNotLoaded();
+		return children.size();
+	}
+
+	/**
+	 * Add a node
+	 * 
+	 * @param node
+	 *            Node
+	 */
+	public void add(FilesystemTreeNode node) {
+		children.add(node);
+	}
+
+	/**
+	 * Load the children to a tmp attribute, **this is the only one method which is called from another thread**
```

Comment:
From which Thread is this method called? And on which Thread is the rest of the class running?